cunumeric/array.py

```diff
@@ -1008,31 +1003,36 @@ def choose(self, choices, out=None, mode="raise", stacklevel=2):
         for c in choices:
             out_shape = broadcast_shapes(out_shape, c.shape)
 
-        # if there is not output specified, create one
-        if out is None:
-            # no output, create one
-            out = ndarray(
-                shape=out_shape,
-                dtype=ch_dtype,
-                stacklevel=stacklevel + 1,
-                inputs=(a, choices),
-            )
-        else:
+        if out is not None and out.dtype == ch_dtype:
             # if output is provided, it shape should be the same as out_shape
             if out.shape != out_shape:
```

Comment:
Make sure you perform the `out.shape` check whenever `out is not None`, even if `out.dtype != ch_dtype`.